File: ruby/lexer_test.rb
require "test/unit"

require_relative "lexer"

class TestLexer< Test::Unit::TestCase
  def test_comment
    lexer = Lexer.new("a #comment \n b")
    t = lexer.next_token
    assert_equal(t.value, "a")
    t = lexer.next_token
    assert_equal(t.value, "b")
  end

  def test_next_token_float_variable
    ("a".."h").each do |var|
      t = Lexer.new(var).next_token
      assert_equal(t.type, TokenType::Variable)
      assert_equal(t.value, var)
      assert_equal(t.var_type, VarType::Float)
    end
  end

  def test_next_token_int_variable
    ("i".."n").each do |var|
      t = Lexer.new(var).next_token
      assert_equal(t.type, TokenType::Variable)
      assert_equal(t.value, var)
      assert_equal(t.var_type, VarType::Int)
    end
  end

  def test_next_token_string_variable
    ("s".."z").each do |var|
      t = Lexer.new(var).next_token
      assert_equal(t.type, TokenType::Variable)
      assert_equal(t.value, var)
      assert_equal(t.var_type, VarType::String)
    end
  end

  def test_next_token_keyword
    KEYWORDS.each do |kw|
      t = Lexer.new(kw.to_s).next_token
      assert_equal(t.type, TokenType::Keyword)
      assert_equal(t.keyword, kw)
    end
  end

  def test_next_token_int_constant
    lexer = Lexer.new("1 234")
    t = lexer.next_token
    assert_equal(t.type, TokenType::Constant)
    assert_equal(t.value, "1")
    assert_equal(t.var_type, VarType::Int)
    t = lexer.next_token
    assert_equal(t.type, TokenType::Constant)
    assert_equal(t.value, "234")
    assert_equal(t.var_type, VarType::Int)
  end

  def test_next_token_float_constant
    lexer = Lexer.new("1.0 234.5")
    t = lexer.next_token
    assert_equal(t.type, TokenType::Constant)
    assert_equal(t.value, "1.0")
    assert_equal(t.var_type, VarType::Float)
    t = lexer.next_token
    assert_equal(t.type, TokenType::Constant)
    assert_equal(t.value, "234.5")
    assert_equal(t.var_type, VarType::Float)
  end

  def test_next_token_string_constant
    lexer = Lexer.new('"a" "abcde"')
    t = lexer.next_token
    assert_equal(t.type, TokenType::Constant)
    assert_equal(t.value, "a")
    assert_equal(t.var_type, VarType::String)
    t = lexer.next_token
    assert_equal(t.type, TokenType::Constant)
    assert_equal(t.value, "abcde")
    assert_equal(t.var_type, VarType::String)
  end

  def test_next_token_symbol
    lexer = Lexer.new(SYMBOL_STRINGS.join(' '))
    for i in 0..10
      t = lexer.next_token
      assert_equal(t.type, TokenType::Symbol)
      assert_equal(t.value, SYMBOL_STRINGS[i])
      assert_equal(t.symbol, SYMBOLS[i])
    end
  end
end

File: python/lexer_test.py
from ssl_token import *
from lexer import Lexer

import unittest


class LexerTest(unittest.TestCase):
    def test_empty(self):
        lexer = Lexer("")
        token = lexer.nextToken()
        self.assertTrue(token.isEof())

    def test_comment(self):
        lexer = Lexer("# nothing\n")
        token = lexer.nextToken()
        self.assertTrue(token.isEof())

    def test_skips_comments(self):
        lexer = Lexer('"hi" # comment\n123\n')
        token = lexer.nextToken()
        self.assertEqual(token.tokenType, TokenType.CONST)
        self.assertEqual(token.varType, VarType.STR)
        self.assertEqual(token.value, "hi")
        token = lexer.nextToken()
        self.assertEqual(token.tokenType, TokenType.CONST)
        self.assertEqual(token.varType, VarType.INT)
        self.assertEqual(token.value, 123)
        token = lexer.nextToken()
        self.assertTrue(token.isEof())

    def test_number(self):
        lexer = Lexer("123")
        token = lexer.nextToken()
        self.assertEqual(token.tokenType, TokenType.CONST)
        self.assertEqual(token.varType, VarType.INT)
        self.assertEqual(token.value, 123)

    def test_float(self):
        lexer = Lexer("123.456")
        token = lexer.nextToken()
        self.assertEqual(token.tokenType, TokenType.CONST)
        self.assertEqual(token.varType, VarType.FLOAT)
        self.assertEqual(token.value, 123.456)

    def test_string_constant(self):
        lexer = Lexer('"123"')
        token = lexer.nextToken()
        self.assertEqual(token.tokenType, TokenType.CONST)
        self.assertEqual(token.varType, VarType.STR)
        self.assertEqual(token.value, "123")

    def test_variables(self):
        expected = [("a", VarType.FLOAT), ("i", VarType.INT),
                    ("s", VarType.STR), ("z", VarType.STR)]
        program = " ".join(value for (value, varType) in expected)
        lexer = Lexer(program)
        for (value, varType) in expected:
            token = lexer.nextToken()
            self.assertEqual(token.tokenType, TokenType.VAR)
            self.assertEqual(token.value, value)
            self.assertEqual(token.varType, varType)

    def test_upper_variables(self):
        expected = [("A", VarType.FLOAT), ("I", VarType.INT),
                    ("S", VarType.STR), ("Z", VarType.STR)]
        program = " ".join(value for (value, varType) in expected)
        lexer = Lexer(program)
        for (value, varType) in expected:
            token = lexer.nextToken()
            self.assertEqual(token.tokenType, TokenType.VAR)
            self.assertEqual(token.value, value)
            self.assertEqual(token.varType, varType)

    def test_keywords(self):
        lexer = Lexer("if then else endif for to step endfor print println")
        expected = [Keyword.IF, Keyword.THEN, Keyword.ELSE, Keyword.ENDIF,
                    Keyword.FOR, Keyword.TO, Keyword.STEP, Keyword.ENDFOR,
                    Keyword.PRINT, Keyword.PRINTLN]
        for keyword in expected:
            token = lexer.nextToken()
            self.assertEqual(token.tokenType, TokenType.KEYWORD)
            self.assertEqual(token.value, keyword)

    def test_upper_keywords(self):
        lexer = Lexer("IF THEN ELSE ENDIF FOR TO STEP ENDFOR PRINT PRINTLN")
        expected = [Keyword.IF, Keyword.THEN, Keyword.ELSE, Keyword.ENDIF,
                    Keyword.FOR, Keyword.TO, Keyword.STEP, Keyword.ENDFOR,
                    Keyword.PRINT, Keyword.PRINTLN]
        for keyword in expected:
            token = lexer.nextToken()
            self.assertEqual(token.tokenType, TokenType.KEYWORD)
            self.assertEqual(token.value, keyword)

    def test_symbols(self):
        symbols = ["+", "-", "*", "/", "=", "<", "<=", "==", "!=", ">=", ">"]
        program = " ".join(symbols)
        lexer = Lexer(program)
        for symbol in symbols:
            token = lexer.nextToken()
            self.assertEqual(token.tokenType, TokenType.SYMBOL)
            self.assertEqual(token.value, symbol)

    def test_symbols_no_spaces(self):
        expected = ["+", "-", "*", "/", "=", "<", "<=", "==", "!=", ">=", ">"]
        lexer = Lexer("+-*/=<<===!=>=>")
        for symbol in expected:
            token = lexer.nextToken()
            self.assertEqual(token.tokenType, TokenType.SYMBOL)
            self.assertEqual(token.value, symbol)


if __name__ == "__main__":
    unittest.main()
